File: Excel/480 Challenge.py
import pandas as pd
import numpy as np

path = "480 Caesar's Cipher_Decrypter.xlsx"
input = pd.read_excel(path, usecols="A:B")
test = pd.read_excel(path, usecols="C")

def decrypt_caesar(encrypted_text, shift):
    def shift_char(char, shift_value):
        if char.isalpha():
            base = 97 if char.islower() else 65
            char_val = ord(char) - base
            shifted_val = (char_val - shift_value) % 26
            return chr(shifted_val + base)
        else:
            return char

    decrypt_char = np.vectorize(shift_char)
    decrypted_text = ''.join(decrypt_char(list(encrypted_text), np.arange(len(encrypted_text)) + shift))
    return decrypted_text

result = input.copy() 
result['Answer Expected'] = result.apply(lambda row: decrypt_caesar(row['Encrypted Text'], row['Shift']), axis=1)

print(result['Answer Expected'].equals(test['Answer Expected'])) # True

File: Excel/400-499/480/480 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/480 Caesar's Cipher_Decrypter.xlsx"
input = read_excel(path, range = "A1:B10")
test  = read_excel(path, range = "C1:C10")

decrypt_caesar <- function(encrypted_text, shift) {
  shift_char <- function(char, shift_value) {
    if (char %in% letters) {
      base <- 97
      char_val <- utf8ToInt(char) - base
      shifted_val <- (char_val - shift_value) %% 26
      intToUtf8(shifted_val + base)
    } else if (char %in% LETTERS) {
      base <- 65
      char_val <- utf8ToInt(char) - base
      shifted_val <- (char_val - shift_value) %% 26
      intToUtf8(shifted_val + base)
    } else {
      char
    }
  }
  decrypt_char <- Vectorize(shift_char, "char")
  decrypted_text <- map2_chr(str_split(encrypted_text, "")[[1]], 0:(nchar(encrypted_text) - 1), 
                             ~ decrypt_char(.x, shift + .y))
  paste0(decrypted_text, collapse = "")
}

result = input %>%
  mutate(`Answer Expected` = map2_chr(`Encrypted Text`, Shift, decrypt_caesar)) %>%
  select(`Answer Expected`)

identical(result, test)
# [1] TRUE